### DISCUSSION TITLE
[Usage question] - Quiescent period start and stop times per trial

### DISCUSSION THREAD

**Discussion by njmiska** [int-brain-lab/iblenv#201] [Q&A]
Hi, question: from what I understand about our task, there is a random amount of time (something like 200-400ms) generated for each trial in which the animal is required to hold the wheel still before the goCue is presented. do you know where that time is stored for each trial?  I've looked in the trials object, but don't seem to see it, maybe it can be calculated using other values within the trials object?

alf.io.load_object(ses_path, object = "trials", namespace = "ibl")

**Comment by nbonacchi** [comment]
Hey Nate,
The specific time of the quiescent period is a value drawn from a truncated exponential between 0.2 and 0.5 with a scale of 0.35 (scale is the inverse of rate), this value is added to 0.2 fixed delay.
This number defines the minimum amount of time that the animal has to maintain wheel movement below a certain threshold and it's a different value every trial. However, if movement is detected, the clock is restarted. So from the point of view of the animal from stim OFF till stim ON nothing is happening in the box, just a gray screen, and there is no way to "know" that the first 0.5 seconds of nothing is actually called an ITI, then there is a random delay of time we call "dead time" where the PC is computing stuff and plotting stuff, then there is the quiescent period start that will only stop if there is no movement for X sec.

Given all this do you want:
- the value of the enforced quiescence that is drawn at each trial; (this is in the raw data and not extracted yet)
- the interval between when this enforcement starts until it ends in a stimON/tone; (in our backlog to extract as the amount of times the clock was restarted)
- the actual time the mouse was exposed to nothing i.e. between stimOFF(t-1) and stimON(t)? (you should be able to get this from the stimOn_times and stimOff_times)

**Comment by njmiska** [comment]
Hey Nicco,

Thanks for this response.  Either the first or (even better) the second would be what I'm looking for.  Basically the time from which the mouse is forced to hold the wheel still until the stim (or trial) starts.

I think I could get an estimate of this for now by using the time of last wheel movement?  Eventually we will probably need to know the values of the 'enforced quiescence' though.

**Comment by nbonacchi** [comment]
_ibl_trials.quiescentValues.npy
_ibl_trials.quiescent_intervals.npy
TODO: dstypes to be approved

**Comment by nbonacchi** [comment]
dataset approved:
_ibl_trials.quiescentValues	[ntrials]	Value of minimum enforced quiescence before stimulus is allowed to appear

regarding quiescent intervals, feedback from DAWG: if the goal is to know the total amount of time the subject was in the pre stim phase, users can get that info from trials.intervals[:,0] and trials.stimOn_times.

**Comment by njmiska** [comment]
Hey Nicco, the goal is not to know the total amount of time the subject was in the pre stim phase, but rather to determine the period of time before each trial in which quiescence was enforced (and the mouse was not moving the wheel).  This is presumably the only time in the task that we can reliably/consistently say neural signals aren't attributable to wheel movement.

Thanks for the update.  Do you have an estimate for when this might be implemented?

**Comment by nbonacchi** [comment]
If you want a quick way of checking from the raw data in ephys sessions before looking for the file:
 
1. Figure out the pregenerated session number of the ephys session from settings file
2. Figure out session trial length from data file
3. Load the corresponding pcqs numpy file
4. there are ~2000 values in the pre generated session: trim it to nTrials 
pregenerated values include `['position', 'quiescence', 'phase', 'probabilityLeft', 'contrastRight', 'contrastLeft']`

Luckily I've already implemented this in the Contrast and probaboility extractor and you can reuse the code easily.

This should do the trick for a given session_path:

```python
import ibllib.io.extractors.ephys_fpga as ephys_fpga
import ibllib.io.raw_data_loaders as raw

session_path = "PATH_TO/FlatIron/mrsicflogellab/Subjects/SWC_054/2020-10-10/001"
data = raw.load_data(session_path)
settings = raw.load_settings(session_path)
pregen_data = ephys_fpga.ProbaContrasts.get_pregenerated_events(data, settings)
pregen_data.keys()
# dict_keys(['position', 'quiescence', 'phase', 'probabilityLeft', 'contrastRight', 'contrastLeft'])
pregen_data['quiescence']
```

**Comment by njmiska** [comment]
Hey Nicco,

To follow up on this, I tried using the code you provided but getting an error "AttributeError: type object 'ProbaContrasts' has no attribute 'get_pregenerated_events'".

I downloaded the raw_behavior_data using code that mayo suggested:
dsets = one.alyx.rest('datasets', 'list', django=f'session,{eid},collection__icontains,raw_behavior_data')
files = one.download_datasets(dsets)

Then ran the code you suggested.  Any idea what the issue might be?  Also, is there a way I can restrict dsets so that it only downloads relevant files? (what it downloaded for this session was pretty large.  eid = '54238fd6-d2d0-4408-b1a9-d19d24fd29ce')

**Comment by nbonacchi** [comment]
Hey Nate :)
looks like you need to update ibllib. If you installed iblenv as per the README just pull from master.

**Comment by njmiska** [comment]
Ah, I see.  Well, that's an issue for now as I'm using an older branch of ibllib (guido-WIP) for dome decoding analysis.  Is there maybe some file(s) I could copy over from master to fix the issue temporarily?

Also, if I read what you said in this thread initially correctly, given that there's a 0.2s fixed delay and at least 0.2s quiescent period before each trial start, is it correct to assume there is at least 400ms quiescent period before each trial?  Is quiescence enforced during the fixed delay period as well or no?  Ie, is it possible that mice move during the fixed delay period and the trial starts anyway?

**Comment by nbonacchi** [comment]
You have to ask guido-WIP to keep the branch up to date with deveop by either fastforwarding it if possible or merging develop in his branch...

Answering your other questions: Yes, Yes, No

**Comment by njmiska** [comment]
great, thanks Nicco
